File: sdks/remote_plugin/python/app/service/session.py
import asyncio
import contextvars
import logging
import time
import uuid
from contextlib import suppress

from app.core import AsyncPlugin, Flow
from app.proto import plugin_pb2 as pb
from app.schema import Payload, Signal

logger = logging.getLogger(__name__)

DEFAULT_ACK_INTERVAL_SEC = 10.0

current_correlation_id: contextvars.ContextVar[str] = contextvars.ContextVar(
    "current_correlation_id",
    default="",
)


class PluginSession(Flow):
    def __init__(
        self,
        instance_id: str,
        plugin: AsyncPlugin,
        output: asyncio.Queue[pb.RuntimeMessage | None],
        *,
        ack_interval_sec: float = DEFAULT_ACK_INTERVAL_SEC,
    ) -> None:
        self.instance_id = instance_id
        self.plugin = plugin
        self.output = output
        self.ack_interval_sec = ack_interval_sec
        self._closed = False
        self._close_lock = asyncio.Lock()
        self._tasks: dict[str, asyncio.Task[None]] = {}

    async def send_payload(
        self,
        payload: Payload,
        *,
        port: int = 0,
    ) -> None:
        msg = self._new_runtime_message("RUNTIME_MESSAGE_TYPE_EMIT_PAYLOAD")
        msg.emit_payload.payload.name = payload.name
        msg.emit_payload.payload.properties = payload.to_json_bytes()
        msg.emit_payload.port = port
        await self.output.put(msg)

    async def send_signal(
        self,
        signal: Signal,
        *,
        port: int = 0,
    ) -> None:
        msg = self._new_runtime_message("RUNTIME_MESSAGE_TYPE_EMIT_SIGNAL")
        msg.emit_signal.signal.name = signal.name
        msg.emit_signal.signal.properties = signal.to_json_bytes()
        msg.emit_signal.port = port
        await self.output.put(msg)

    def _new_runtime_message(
        self,
        message_type: str,
        *,
        correlation_id: str | None = None,
    ) -> pb.RuntimeMessage:
        return pb.RuntimeMessage(
            instance_id=self.instance_id,
            message_id=uuid.uuid4().hex,
            correlation_id=(
                current_correlation_id.get() if correlation_id is None else correlation_id
            ),
            type=message_type,
        )

    async def process_stream(self, request_iterator) -> None:
        try:
            async for message in request_iterator:
                if message.type == pb.RUNTIME_MESSAGE_TYPE_CANCEL:
                    self._handle_cancel(message)
                    continue
                self._start_message_task(message)
        except asyncio.CancelledError:
            await self.close()
            raise
        except Exception:
            logger.exception("remote plugin runtime stream failed")
        finally:
            await self._drain_running_tasks()
            with suppress(Exception):
                await self.close()

    async def close(self) -> None:
        async with self._close_lock:
            if self._closed:
                return
            self._closed = True
            for task in self._tasks.values():
                task.cancel()
            await self.output.put(None)

    def _start_message_task(self, message: pb.RuntimeMessage) -> None:
        correlation_id = message.message_id
        task = asyncio.create_task(self._handle_runtime_message(message))
        self._tasks[correlation_id] = task

    def _handle_cancel(self, message: pb.RuntimeMessage) -> None:
        correlation_id = message.correlation_id
        if not correlation_id:
            return
        task = self._tasks.get(correlation_id)
        if task is not None and not task.done():
            task.cancel()

    async def _handle_runtime_message(self, message: pb.RuntimeMessage) -> None:
        correlation_id = message.message_id
        token = current_correlation_id.set(correlation_id)
        ack_task: asyncio.Task[None] | None = None
        try:
            await self._send_ack(correlation_id)
            ack_task = asyncio.create_task(self._ack_renew_loop(correlation_id))

            await self._dispatch_message(message)
        except asyncio.CancelledError:
            report = self._new_runtime_message(
                "RUNTIME_MESSAGE_TYPE_REPORT",
                correlation_id=correlation_id,
            )
            report.report.status = pb.REPORT_STATUS_CANCELED
            await self.output.put(report)
        except Exception as exc:
            logger.exception(
                "remote plugin event failed instance_id=%s message_id=%s type=%s",
                message.instance_id,
                message.message_id,
                message.type,
            )
            report = self._new_runtime_message("RUNTIME_MESSAGE_TYPE_REPORT")
            report.report.status = pb.REPORT_STATUS_ERROR
            report.report.error.code = exc.__class__.__name__
            report.report.error.message = str(exc)
            await self.output.put(report)
        else:
            report = self._new_runtime_message("RUNTIME_MESSAGE_TYPE_REPORT")
            report.report.status = pb.REPORT_STATUS_OK
            await self.output.put(report)
        finally:
            if ack_task is not None:
                ack_task.cancel()
                with suppress(asyncio.CancelledError):
                    await ack_task
            current_correlation_id.reset(token)
            self._tasks.pop(correlation_id, None)

    async def _send_ack(self, correlation_id: str) -> None:
        ack = self._new_runtime_message(
            "RUNTIME_MESSAGE_TYPE_ACK",
            correlation_id=correlation_id,
        )
        ack.ack.timestamp = time.time_ns() // 1_000_000
        await self.output.put(ack)

    async def _ack_renew_loop(self, correlation_id: str) -> None:
        while True:
            await asyncio.sleep(self.ack_interval_sec)
            await self._send_ack(correlation_id)

    async def _dispatch_message(self, message: pb.RuntimeMessage) -> None:
        body = message.WhichOneof("body")
        match body:
            case "lifecycle":
                await self._dispatch_lifecycle(message.lifecycle.type)
            case "signal":
                await self.plugin.on_signal(
                    self,
                    Signal.from_json_bytes(message.signal.name, message.signal.properties),
                )
            case "payload":
                await self.plugin.on_payload(
                    self,
                    Payload.from_json_bytes(message.payload.name, message.payload.properties),
                )
            case _:
                raise ValueError(f"unsupported runtime message body: {body or '<empty>'}")

    async def _dispatch_lifecycle(self, lifecycle_type: int) -> None:
        match lifecycle_type:
            case pb.LIFECYCLE_TYPE_START:
                await self.plugin.on_start(self)
            case pb.LIFECYCLE_TYPE_READY:
                await self.plugin.on_ready(self)
            case pb.LIFECYCLE_TYPE_PAUSE:
                await self.plugin.on_pause()
            case pb.LIFECYCLE_TYPE_RESUME:
                await self.plugin.on_resume(self)
            case pb.LIFECYCLE_TYPE_STOP:
                await self.plugin.on_stop()
            case _:
                raise ValueError(f"unsupported lifecycle type: {lifecycle_type}")

    async def _drain_running_tasks(self) -> None:
        tasks = list(self._tasks.values())
        if not tasks:
            return
        await asyncio.gather(*tasks, return_exceptions=True)
        self._tasks.clear()

File: sdks/remote_plugin/nodejs/src/service/session.ts
import { AsyncLocalStorage } from 'node:async_hooks'
import type { Flow } from '../core/flow.js'
import type { AsyncPlugin } from '../core/plugin.js'
import { Signal } from '../schema/signal.js'
import { Payload } from '../schema/payload.js'
import { AsyncQueue } from '../utils/async-queue.js'
import {
  LifecycleType,
  ReportStatus,
  RuntimeMessageType,
  type RuntimeMessage,
} from '../proto/plugin.js'

// ---------------------------------------------------------------------------
// Correlation ID context (AsyncLocalStorage)
// ---------------------------------------------------------------------------

const correlationIdStorage = new AsyncLocalStorage<string>()

export function getCurrentCorrelationId(): string {
  return correlationIdStorage.getStore() ?? ''
}

// ---------------------------------------------------------------------------
// PluginSession
// ---------------------------------------------------------------------------

/**
 * Manages a single RunInstance bidirectional stream for one plugin instance.
 *
 * Implements the Flow interface so that plugins can emit payloads/signals
 * directly through the session's outgoing message queue.
 */
export class PluginSession implements Flow {
  readonly instanceId: string
  readonly outgoingMessages: AsyncQueue<RuntimeMessage>

  private readonly plugin: AsyncPlugin<unknown>
  private readonly ackIntervalMs: number
  private closed = false
  private readonly runningTasks = new Map<string, AbortController>()

  constructor(
    instanceId: string,
    plugin: AsyncPlugin<unknown>,
    outgoingMessages: AsyncQueue<RuntimeMessage>,
    options: { ackIntervalSec: number },
  ) {
    this.instanceId = instanceId
    this.plugin = plugin
    this.outgoingMessages = outgoingMessages
    this.ackIntervalMs = options.ackIntervalSec * 1000
  }

  // ── Flow interface ───────────────────────────────────────────────────────

  async sendPayload(payload: Payload, options?: { port?: number }): Promise<void> {
    const msg = this.newRuntimeMessage(RuntimeMessageType.RUNTIME_MESSAGE_TYPE_EMIT_PAYLOAD)
    msg.body = {
      $case: 'emitPayload' as const,
      emitPayload: {
        payload: {
          name: payload.name,
          properties: payload.toJsonBytes(),
        },
        port: options?.port ?? 0,
      },
    }
    this.outgoingMessages.put(msg)
  }

  async sendSignal(signal: Signal, options?: { port?: number }): Promise<void> {
    const msg = this.newRuntimeMessage(RuntimeMessageType.RUNTIME_MESSAGE_TYPE_EMIT_SIGNAL)
    msg.body = {
      $case: 'emitSignal' as const,
      emitSignal: {
        signal: {
          name: signal.name,
          properties: signal.toJsonBytes(),
        },
        port: options?.port ?? 0,
      },
    }
    this.outgoingMessages.put(msg)
  }

  // ── Stream processing ────────────────────────────────────────────────────

  /**
   * Reads incoming RuntimeMessages from the gRPC stream and dispatches
   * them to the plugin. Each message is handled in its own "task" (Promise)
   * tracked by correlation_id, allowing cancel propagation.
   */
  async processStream(messages: AsyncIterable<RuntimeMessage>): Promise<void> {
    try {
      for await (const message of messages) {
        if (message.type === RuntimeMessageType.RUNTIME_MESSAGE_TYPE_CANCEL) {
          this.handleCancel(message)
          continue
        }
        this.startMessageTask(message)
      }
    } catch (err) {
      if (!isAbortError(err)) {
        console.error(
          `remote plugin runtime stream failed instance_id=${this.instanceId}`,
          err,
        )
      }
    } finally {
      await this.drainRunningTasks()
      await this.close()
    }
  }

  async close(): Promise<void> {
    if (this.closed) return
    this.closed = true

    // Cancel all running tasks
    for (const ac of this.runningTasks.values()) {
      ac.abort()
    }
    this.runningTasks.clear()

    // Signal the output queue to terminate
    this.outgoingMessages.put(null)
  }

  // ── private: task management ─────────────────────────────────────────────

  private startMessageTask(message: RuntimeMessage): void {
    const correlationId = message.messageId
    const ac = new AbortController()
    this.runningTasks.set(correlationId, ac)

    this.handleRuntimeMessage(message, ac.signal)
      .catch(() => {}) // errors handled inside handleRuntimeMessage
      .finally(() => {
        this.runningTasks.delete(correlationId)
      })
  }

  private handleCancel(message: RuntimeMessage): void {
    const correlationId = message.correlationId
    if (!correlationId) return
    const ac = this.runningTasks.get(correlationId)
    ac?.abort()
  }

  private async handleRuntimeMessage(
    message: RuntimeMessage,
    signal: AbortSignal,
  ): Promise<void> {
    const correlationId = message.messageId
    let ackTimer: ReturnType<typeof setInterval> | undefined

    try {
      // Send initial ACK
      this.sendAck(correlationId)

      // Start ACK keepalive
      ackTimer = setInterval(() => this.sendAck(correlationId), this.ackIntervalMs)

      // Run dispatch in correlation context
      await correlationIdStorage.run(correlationId, () =>
        this.dispatchMessage(message, signal),
      )

      // Success report
      this.sendReport(correlationId, ReportStatus.REPORT_STATUS_OK)
    } catch (err) {
      if (isAbortError(err)) {
        console.info(
          `remote plugin task canceled instance_id=${this.instanceId} correlation_id=${correlationId}`,
        )
        this.sendReport(correlationId, ReportStatus.REPORT_STATUS_CANCELED)
      } else {
        console.error(
          `remote plugin event failed instance_id=${this.instanceId} message_id=${message.messageId} type=${message.type}`,
          err,
        )
        this.sendReport(
          correlationId,
          ReportStatus.REPORT_STATUS_ERROR,
          err instanceof Error ? err : new Error(String(err)),
        )
      }
    } finally {
      if (ackTimer) clearInterval(ackTimer)
    }
  }

  private async dispatchMessage(
    message: RuntimeMessage,
    signal: AbortSignal,
  ): Promise<void> {
    const body = message.body
    if (!body) throw new Error('runtime message body is empty')

    // Check abort before dispatching
    throwIfAborted(signal)

    switch (body.$case) {
      case 'lifecycle':
        await this.dispatchLifecycle(body.lifecycle.type)
        break
      case 'signal':
        await this.plugin.onSignal(
          this,
          Signal.fromJsonBytes(body.signal.name, body.signal.properties),
        )
        break
      case 'payload':
        await this.plugin.onPayload(
          this,
          Payload.fromJsonBytes(body.payload.name, body.payload.properties),
        )
        break
      default:
        throw new Error(`unsupported runtime message body: ${(body as { $case: string }).$case}`)
    }
  }

  private async dispatchLifecycle(type: LifecycleType): Promise<void> {
    switch (type) {
      case LifecycleType.LIFECYCLE_TYPE_START:
        await this.plugin.onStart(this)
        break
      case LifecycleType.LIFECYCLE_TYPE_READY:
        await this.plugin.onReady(this)
        break
      case LifecycleType.LIFECYCLE_TYPE_PAUSE:
        await this.plugin.onPause()
        break
      case LifecycleType.LIFECYCLE_TYPE_RESUME:
        await this.plugin.onResume(this)
        break
      case LifecycleType.LIFECYCLE_TYPE_STOP:
        await this.plugin.onStop()
        break
      default:
        throw new Error(`unsupported lifecycle type: ${type}`)
    }
  }

  // ── private: message builders ────────────────────────────────────────────

  private newRuntimeMessage(type: RuntimeMessageType, correlationId?: string): RuntimeMessage {
    return {
      instanceId: this.instanceId,
      messageId: crypto.randomUUID().replace(/-/g, ''),
      correlationId: correlationId ?? getCurrentCorrelationId(),
      type,
      metadata: {},
      body: undefined,
    }
  }

  private sendAck(correlationId: string): void {
    const msg = this.newRuntimeMessage(RuntimeMessageType.RUNTIME_MESSAGE_TYPE_ACK, correlationId)
    msg.body = {
      $case: 'ack' as const,
      ack: { timestamp: Date.now() },
    }
    this.outgoingMessages.put(msg)
  }

  private sendReport(
    correlationId: string,
    status: ReportStatus,
    error?: Error,
  ): void {
    const msg = this.newRuntimeMessage(
      RuntimeMessageType.RUNTIME_MESSAGE_TYPE_REPORT,
      correlationId,
    )
    msg.body = {
      $case: 'report' as const,
      report: {
        status,
        error: error
          ? { code: error.constructor.name, message: error.message, details: '' }
          : undefined,
      },
    }
    this.outgoingMessages.put(msg)
  }

  // ── private: cleanup ─────────────────────────────────────────────────────

  private async drainRunningTasks(): Promise<void> {
    const tasks: Promise<void>[] = []
    for (const [, ac] of this.runningTasks) {
      ac.abort()
    }
    // Wait for all tasks to settle
    await Promise.allSettled(tasks)
    this.runningTasks.clear()
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function isAbortError(err: unknown): boolean {
  return err instanceof DOMException && err.name === 'AbortError'
    || err instanceof Error && err.name === 'AbortError'
}

function throwIfAborted(signal: AbortSignal): void {
  if (signal.aborted) {
    throw new DOMException('The operation was aborted', 'AbortError')
  }
}
